src/hotspot/share/gc/serial/serialHeap.cpp

```diff
@@ -762,12 +761,40 @@ bool SerialHeap::no_allocs_since_save_marks() {
          _old_gen->no_allocs_since_save_marks();
 }
 
+template <typename OopClosureType>
+static void oop_iterate_from(OopClosureType* blk, ContiguousSpace* space, HeapWord* from) {
+  assert(from != nullptr, "precondition");
+  HeapWord* t;
+  HeapWord* p = from;
+
+  const intx interval = PrefetchScanIntervalInBytes;
+  do {
+    t = space->top();
+    while (p < t) {
+      Prefetch::write(p, interval);
+      oop m = cast_to_oop(p);
+      p += m->oop_iterate_size(blk);
+    }
+  } while (t < space->top());
+}
+
 void SerialHeap::scan_evacuated_objs(YoungGenScanClosure* young_cl,
                                      OldGenScanClosure* old_cl) {
   do {
-    young_gen()->oop_since_save_marks_iterate(young_cl);
-    old_gen()->oop_since_save_marks_iterate(old_cl);
-  } while (!no_allocs_since_save_marks());
+    {
+      HeapWord* saved_top = young_gen()->saved_mark_word();
+      oop_iterate_from(young_cl, young_gen()->to(), saved_top);
+      young_gen()->set_saved_mark_word();
+    }
+    {
+      HeapWord* saved_top = old_gen()->saved_mark_word();
+      oop_iterate_from(old_cl, old_gen()->space(), saved_top);
+      old_gen()->set_saved_mark_word();
+    }
+    if (young_gen()->no_allocs_since_save_marks()) {
+      break;
+    }
+  } while (true);
```

Comment:
Why not keeping the previous ` while (!no_allocs_since_save_marks())`?

Reply:
The old-gen-iteration should have ensured no alloc in old-gen, so only young-gen requires rechecking.

Reply:
Both new-gen `YoungGenScanClosure` and old -gen `OldGenScanClosure` use the methods `ScavengeHelper::try_scavenge` and `DefNewGeneration::copy_to_survivor_space` to evacuate objects. The `DefNewGeneration::copy_to_survivor_space` may promote objects to old-gen. Do I miss anything or misunderstand it?

Reply:
Both closures can advance `top`s in young/old gen. However, the postcondition of `oop_iterate_from` is all newly evacuated objs in the corresponding space are scanned.

Reply:
Ohh, I see. The old-gen is scaned finally, so the objects in old-gen are totally processed. But the to-space of the young-gen is firstly scaned, so we need to judge it.

What about `while (!young_gen()->no_allocs_since_save_marks())`? But it is OK now.

Reply:
That should also work; identical asm I assume.

(The double negation (`!` and `no_`) makes it slightly less readable, so I went for this `break` approach.)

Reply:
Would it be possible to rename the method then? The `break` approach is much uglier.